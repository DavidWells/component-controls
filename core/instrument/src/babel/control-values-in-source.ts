import {
  Story,
  StoryArguments,
  StoryArgument,
} from '@component-controls/specification';
import { adjustSourceLocation } from './utils';

const findArguments = (
  args: StoryArguments,
  paramName: string,
): StoryArgument | undefined => {
  let result: StoryArgument | undefined;
  for (let i = 0; i < args.length; i += 1) {
    const p = args[i];
    if (Array.isArray(p.value)) {
      result = findArguments(p.value, paramName);
      if (result) {
        break;
      }
    } else {
      if (p.value === paramName) {
        result = p;
        break;
      }
    }
  }
  return result;
};
export const addArgumentUsage = (
  story: Story,
  args: StoryArguments,
  node: any,
) => {
  const param = findArguments(args, node.name);

  if (param) {
    if (param.usage === undefined) {
      param.usage = [];
    }
    param.usage.push(adjustSourceLocation(story, node.loc));
  }
};
export const extractArgumentsUsage = (story: Story, args: StoryArguments) => {
  return {
    Identifier: (path: any) => {
      addArgumentUsage(story, args, path.node);
    },
  };
};
