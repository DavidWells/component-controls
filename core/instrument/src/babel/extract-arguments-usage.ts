import {
  Story,
  StoryArguments,
  StoryArgument,
  ArgUsageLocation,
} from '@component-controls/specification';
import { adjustSourceLocation } from './utils';

const findArguments = (
  args: StoryArguments,
  paramName: string,
): StoryArgument | undefined => {
  let result: StoryArgument | undefined;
  for (let i = 0; i < args.length; i += 1) {
    const p = args[i];
    if (Array.isArray(p.value)) {
      result = findArguments(p.value, paramName);
      if (result) {
        break;
      }
    } else {
      if (p.value === paramName) {
        result = p;
        break;
      }
    }
  }
  return result;
};
export const addArgumentUsage = (
  story: Story,
  args: StoryArguments,
  node: any,
): ArgUsageLocation | undefined => {
  const param = findArguments(args, node.name);

  if (param) {
    if (param.usage === undefined) {
      param.usage = [];
    }
    const loc = adjustSourceLocation(story, node.loc);
    const existing = param.usage.find(
      p =>
        p.loc.start.line === loc.start.line &&
        p.loc.start.column === loc.start.column &&
        p.loc.end.line === loc.end.line &&
        p.loc.end.column === loc.end.column,
    );
    const usage: ArgUsageLocation = { loc };
    if (!existing) {
      param.usage.push(usage);
      return usage;
    }
  }
  return undefined;
};
export const extractArgumentsUsage = (story: Story, args: StoryArguments) => {
  return {
    TemplateLiteral: (path: any) => {
      path.node.expressions.forEach((expression: any) => {
        if (expression.type === 'Identifier') {
          addArgumentUsage(story, args, expression);
        }
      });
    },
    JSXSpreadAttribute: (path: any) => {
      if (path.node.argument.type === 'Identifier') {
        addArgumentUsage(story, args, path.node.argument);
      }
    },
    Identifier: (path: any) => {
      addArgumentUsage(story, args, path.node);
    },
    Property: (path: any) => {
      const node = path.node;
      if (node.value.type === 'Identifier' && node.key.type === 'Identifier') {
        const usage = addArgumentUsage(story, args, node.value);
        if (usage) {
          usage.name = {
            loc: adjustSourceLocation(story, node.key.loc),
            name: node.key.name,
          };
          path.skip();
        }
      }
    },
  };
};
