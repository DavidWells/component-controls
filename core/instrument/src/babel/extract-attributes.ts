import { StoryParameters } from '@component-controls/specification';

interface StoryAttribute {
  name: string;
  value: any;
}
const nodeToValue = (node: any): any => {
  if (node) {
    switch (node.type) {
      case 'BooleanLiteral':
      case 'NumericLiteral':
      case 'StringLiteral':
        return node.value;
      case 'Identifier':
        return node.name;
      case 'Property':
        return node.name;
      case 'ObjectProperty':
        console.log('ObjectProperty', node.key.value);
        return node.key.value;

      case 'Literal':
        return node.raw;
      case 'RegExpLiteral':
        const value = node.raw ?? node.extra ? node.extra.raw : undefined;
        // remove leading trailing slashes for string to reg xonversion
        return typeof value === 'string'
          ? value.replace(/^\/|\/$/g, '')
          : value;
      case 'MemberExpression':
        return `${node.object.name}.${node.property.name}`;
      case 'ObjectExpression':
        return extractAttributes(node);
      case 'ArrayExpression':
        if (Array.isArray(node.elements)) {
          return node.elements.map((v: any) => nodeToValue(v));
        }
        break;
      default:
        // console.log(property.value);
        return undefined;
    }
  }
  return undefined;
};
const nodeToAttribute = (node: any): StoryAttribute | undefined => {
  const value = node.value || node;
  const name = node.key ? node.key.name ?? node.key.value : node.name;
  const retVal = nodeToValue(value);
  return retVal !== undefined
    ? name
      ? { value: retVal, name }
      : retVal
    : undefined;
};
export const extractAttributes = (
  node: any,
): StoryParameters | any | undefined => {
  if (node) {
    if (node.properties) {
      const attributes: StoryParameters = node.properties.reduce(
        (acc: StoryParameters, propNode: any) => {
          const attribute = nodeToAttribute(propNode);
          if (attribute) {
            return { ...acc, [attribute.name]: attribute.value };
          } else {
            return acc;
          }
        },
        {},
      );
      return attributes;
    }
    const attribute = nodeToAttribute(node);
    if (attribute) {
      return attribute.value;
    }
  }
  return undefined;
};
