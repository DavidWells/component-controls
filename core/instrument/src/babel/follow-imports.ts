import * as fs from 'fs';
import * as resolve from 'resolve';
import * as parser from '@babel/parser';
import * as path from 'path';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { CodeLocation } from '@component-controls/specification';
import { ImportTypes, traverseImports } from './extract-imports';
import {
  traverseExports,
  ExportTypes,
  ExportType,
  EXPORT_ALL,
} from './extract-exports';

export interface FollowImportType {
  exportedAs: string;
  from: string;
  filePath?: string;
  loc?: CodeLocation;
  source?: string;
}

export const followImports = (
  importName: string,
  filePath: string,
  parserOptions?: parser.ParserOptions,
  resolveOptions?: resolve.SyncOpts,
  initialAST?: File,
): FollowImportType | undefined => {
  const source = fs.readFileSync(filePath, 'utf8');
  const ast = initialAST || parser.parse(source, parserOptions);

  const baseImportedName = importName.split('.')[0];

  const exports: ExportTypes = {
    named: {},
  };
  traverse(ast, traverseExports(exports));
  const folderName = path.dirname(filePath);
  const findExport =
    baseImportedName === 'default' || baseImportedName === 'namespace'
      ? exports.default
      : exports.named[baseImportedName];
  if (findExport !== undefined) {
    if (!findExport.from) {
      return {
        filePath,
        exportedAs: findExport.name,
        loc: findExport.loc,
        from: '',
        source,
      };
    } else {
      const resolvedFilePath = resolve.sync(findExport.from, {
        ...resolveOptions,
        basedir: folderName,
      });
      const imported = followImports(
        findExport.internalName,
        resolvedFilePath,
        parserOptions,
        resolveOptions,
      );
      return imported ? { ...imported, from: findExport.from } : undefined;
    }
  }
  const allExports: ExportType[] = Object.keys(exports.named).reduce(
    (acc: ExportType[], key: string) => [...acc, exports.named[key]],
    [],
  );
  let foundInExportAll: FollowImportType | undefined;
  allExports
    .filter(e => e.internalName === EXPORT_ALL && e.from)
    .some(e => {
      if (e.from) {
        const resolvedFilePath = resolve.sync(e.from, {
          ...resolveOptions,
          basedir: folderName,
        });
        foundInExportAll = followImports(
          baseImportedName,
          resolvedFilePath,
          parserOptions,
          resolveOptions,
        );
      }
      return { ...foundInExportAll, from: e.from };
    });
  if (foundInExportAll) {
    return foundInExportAll;
  }
  const imports: ImportTypes = {};
  traverse(ast, traverseImports(imports));
  const findImport = imports[baseImportedName];
  if (findImport) {
    try {
      const resolvedFilePath = resolve.sync(findImport.from, {
        ...resolveOptions,
        basedir: folderName,
      });
      const imported = followImports(
        findImport.importedName,
        resolvedFilePath,
        parserOptions,
        resolveOptions,
      );
      return imported ? { ...imported, from: findImport.from } : undefined;
    } catch (e) {
      //non-exxsiting file
      return { exportedAs: findImport.importedName, from: findImport.from };
    }
  }
  return undefined;
};
