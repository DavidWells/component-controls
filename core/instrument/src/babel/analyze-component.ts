import fs from 'fs';
import * as parser from '@babel/parser';
import traverse, { TraverseOptions } from '@babel/traverse';
import { Component, JSXNode, JSXTree } from '@component-controls/core';
import { collectAttributes } from './extract-attributes';
import { InstrumentOptions } from '../types';

type JSXLinkedNode = JSXNode & { parent?: JSXNode };
type JSXLinkedTree = JSXLinkedNode[];

export const traverseJSX = (jsx: JSXLinkedTree): TraverseOptions => {
  let current: JSXLinkedNode = { children: jsx };
  return {
    JSXElement: {
      enter(path) {
        const node = path.node;
        const name = node.openingElement.name as { name: string };
        const attributes = collectAttributes(node.openingElement);
        const jsxNode: JSXLinkedNode = {
          name: name.name,
          parent: current,
          children: [],
          attributes: Object.keys(attributes),
        };
        if (current.children) {
          current.children.push(jsxNode);
        }
        current = jsxNode;
      },

      exit() {
        current = current.parent as JSXLinkedNode;
      },
    },
  };
};
export const analyze_components = (
  component: Component,
  filePath: string,
  options?: InstrumentOptions,
): void => {
  const { parser: parserOptions } = options || {};
  const source = fs.readFileSync(filePath, 'utf8');
  const ast = parser.parse(source, parserOptions);
  const jsx: JSXLinkedTree = [];
  traverse(ast, traverseJSX(jsx));
  const mapJSXTree = (input?: JSXLinkedTree): JSXTree => {
    return input
      ? // eslint-disable-next-line @typescript-eslint/no-unused-vars
        input.map(({ parent, children, ...rest }) => ({
          children: mapJSXTree(children),
          ...rest,
        }))
      : [];
  };
  component.jsx = mapJSXTree(jsx);
};
