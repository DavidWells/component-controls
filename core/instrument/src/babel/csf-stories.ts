import {
  StoriesStore,
  Story,
  StoriesDoc,
  Stories,
} from '@component-controls/specification';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { extractFunctionParameters } from './extract-function-parameters';
import { extractAttributes } from './extract-attributes';
import { followStoryImport } from './follow-imports';
import { componentsFromParams } from '../misc/component-attributes';
import { sourceLocation } from '../misc/source-location';
import { ParseStorieReturnType, InstrumentOptions } from '../types';

export const extractCSFStories = (
  ast: File,
  _options: InstrumentOptions,
  { source, filePath }: { source: string; filePath: string },
): ParseStorieReturnType => {
  const globals: Stories = {};
  const localStories: Stories = {};

  const extractArrowFunction = (
    path: any,
    declaration: any,
  ): Story | undefined => {
    if (declaration.init) {
      switch (declaration.init.type) {
        case 'ArrowFunctionExpression': {
          const el = declaration.init.body;
          const name = declaration.id.name;
          const story: Story = {
            loc: sourceLocation(el.loc),
            name,
            id: name,
          };
          traverse(path.node, extractFunctionParameters(story), path.scope);
          return story;
        }
        case 'Identifier': {
          return followStoryImport(
            declaration.id.name,
            declaration.init.name,
            filePath,
            source,
            _options,
            ast,
          );
        }
      }
    }
    return undefined;
  };
  let components: { [key: string]: string | undefined } = {};
  const store: StoriesStore = {
    stories: {},
    docs: {},
    components: {},
    packages: {},
  };
  traverse(ast as any, {
    ExportDefaultDeclaration: (path: any) => {
      const { declaration } = path.node;
      const attributes = extractAttributes(declaration);

      const { title } = attributes || {};
      if (typeof title === 'string') {
        const attrComponents = componentsFromParams(attributes);
        components = attrComponents.reduce(
          (acc, componentName) => ({ ...acc, [componentName]: undefined }),
          components,
        );
        const doc: StoriesDoc = {
          ...attributes,
          components: {},
        };
        if (attrComponents.length > 0) {
          doc.component = attrComponents[0];
        }
        store.docs[title] = doc;
      }
    },
    AssignmentExpression: (path: any) => {
      const node = path.node;
      if (
        node.left.type === 'MemberExpression' &&
        node.left.property.type === 'Identifier' &&
        node.left.property.name === 'story' &&
        node.right.type === 'ObjectExpression'
      ) {
        const storyName = node.left.object.name;
        const { name = storyName, ...attributes } = extractAttributes(
          node.right,
        );
        globals[storyName] = {
          ...attributes,
          name,
        };

        if (store.stories[storyName]) {
          const attrComponents = componentsFromParams(attributes);
          components = attrComponents.reduce(
            (acc, componentName) => ({ ...acc, [componentName]: undefined }),
            components,
          );
          const story: Story = {
            ...attributes,
            ...store.stories[storyName],
            name,
          };
          if (attrComponents.length > 0) {
            story.component = attrComponents[0];
          }
          store.stories[storyName] = story;
        }
      }
    },
    VariableDeclaration: (path: any) => {
      const { declarations } = path.node;

      if (Array.isArray(declarations) && declarations.length > 0) {
        const declaration = declarations[0];
        if (declaration) {
          const name = declaration.id.name;
          //check if it was a named export
          if (!store.stories[name]) {
            const story = extractArrowFunction(path, declaration);
            if (story && story.name) {
              localStories[story.name] = story;
            }
          }
        }
      }
    },
    ExportSpecifier: (path: any) => {
      const { node } = path;
      const localName = node.local.name;
      const exportedName = node.exported.name;
      const story = localStories[localName];
      if (story) {
        const global = globals[localName];
        if (global) {
          localStories[localName] = {
            ...story,
            ...global,
          };
        }
        store.stories[exportedName] = story;
      }
    },
    ExportNamedDeclaration: (path: any) => {
      const { declaration } = path.node;
      if (declaration) {
        const { declarations } = declaration;
        if (Array.isArray(declarations) && declarations.length > 0) {
          const story = extractArrowFunction(path, declarations[0]);
          if (story) {
            const name = story.name;
            store.stories[name] = {
              ...story,
              ...globals[name],
            };
          }
        }
      }
    },
  });
  if (Object.keys(store.docs).length === 1) {
    //@ts-ignore
    store.docs[Object.keys(store.docs)[0]].components = components;
  } else {
    throw new Error(`stories should have one default export`);
  }
  return store;
};
