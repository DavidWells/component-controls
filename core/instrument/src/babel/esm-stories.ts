import { Story, Document, Stories } from '@component-controls/core';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { extractFunctionParameters } from './extract-function-parameters';
import { extractAttributes } from './extract-attributes';
import { followStoryImport } from './follow-imports';
import { componentsFromParams } from '../misc/component-attributes';
import { sourceLocation } from '../misc/source-location';
import {
  LoadingDocStore,
  ParseStorieReturnType,
  InstrumentOptions,
} from '../types';

export const extractCSFStories = (
  ast: File,
  _options: InstrumentOptions,
  { source, filePath }: { source: string; filePath: string },
): ParseStorieReturnType => {
  const globals: Stories = {};
  const localFunctions: Stories = {};

  const extractFunction = (path: any, declaration: any): Story | undefined => {
    if (declaration.init) {
      switch (declaration.init.type) {
        case 'ArrowFunctionExpression': {
          const el = declaration.init.body;
          const name = declaration.id.name;
          const story: Story = {
            loc: sourceLocation(el.loc),
            name,
            id: name,
          };
          traverse(path.node, extractFunctionParameters(story), path.scope);
          return story;
        }
        case 'Identifier': {
          return followStoryImport(
            declaration.id.name,
            declaration.init.name,
            filePath,
            source,
            _options,
            ast,
          );
        }
        case 'CallExpression': {
          //template.bind
          if (declaration.init.callee?.property?.name === 'bind') {
            const callee = declaration.init.callee?.object;
            if (callee?.name) {
              const template = localFunctions[callee.name];
              if (template) {
                const name = declaration.id.name;
                const story: Story = {
                  loc: template.loc,
                  name,
                  id: name,
                  arguments: template.arguments,
                };
                return story;
              }
            }
          }
        }
      }
    } else if (declaration.type === 'FunctionDeclaration') {
      const name = declaration.id.name;
      const story: Story = {
        loc: sourceLocation(declaration.body.loc),
        name,
        id: name,
      };
      traverse(path.node, extractFunctionParameters(story), path.scope);
      return story;
    }
    return undefined;
  };
  let components: { [key: string]: string | undefined } = {};
  const store: LoadingDocStore = {
    stories: {},
    doc: undefined,
    components: {},
    packages: {},
  };
  traverse(ast as any, {
    ExportDefaultDeclaration: (path: any) => {
      const { declaration } = path.node;
      const attributes = extractAttributes(declaration);

      const { title: docTitle, name } = attributes || {};
      const title = docTitle || name;
      if (typeof title === 'string') {
        const attrComponents = componentsFromParams(attributes);
        components = attrComponents.reduce(
          (acc, componentName) => ({ ...acc, [componentName]: undefined }),
          components,
        );
        const doc: Document = {
          ...attributes,
          title,
          components: {},
        };
        if (attrComponents.length > 0) {
          doc.component = attrComponents[0];
        }

        store.doc = doc;
      }
    },
    AssignmentExpression: (path: any) => {
      const node = path.node;
      const storyExport = node.left?.object?.name;
      if (
        node.left.type === 'MemberExpression' &&
        node.left.property.type === 'Identifier' &&
        store.stories[storyExport]
      ) {
        const extractedValue = extractAttributes(node.right);
        const extractedProps =
          node.left.property.name === 'story'
            ? extractedValue
            : { [node.left.property.name]: extractedValue };
        const { name, storyName, ...attributes } = extractedProps;

        const nameAttr = storyName || name;
        if (store.stories[storyExport]) {
          const attrComponents = componentsFromParams(attributes);
          components = attrComponents.reduce(
            (acc, componentName) => ({ ...acc, [componentName]: undefined }),
            components,
          );
          const story: Story = {
            ...attributes,
            ...store.stories[storyExport],
          };
          if (nameAttr) {
            story.name = nameAttr;
          }
          store.stories[storyExport] = story;
          globals[storyExport] = story;
        }
      }
    },
    VariableDeclaration: (path: any) => {
      const { declarations } = path.node;

      if (Array.isArray(declarations) && declarations.length > 0) {
        const declaration = declarations[0];
        if (declaration) {
          const name = declaration.id.name;
          //check if it was a named export
          if (!store.stories[name]) {
            const story = extractFunction(path, declaration);
            if (story && story.name) {
              localFunctions[story.name] = story;
            }
          }
        }
      }
    },
    ExportSpecifier: (path: any) => {
      const { node } = path;
      const localName = node.local.name;
      const exportedName = node.exported.name;
      const story = localFunctions[localName];
      if (story) {
        const global = globals[localName];
        if (global) {
          localFunctions[localName] = {
            ...story,
            ...global,
          };
        }
        store.stories[exportedName] = story;
      }
    },
    ExportNamedDeclaration: (path: any) => {
      const { declaration } = path.node;
      if (declaration) {
        const { declarations } = declaration;
        if (declarations) {
          if (Array.isArray(declarations) && declarations.length > 0) {
            const story = extractFunction(path, declarations[0]);
            if (story) {
              const name = story.name;
              store.stories[name] = {
                ...story,
                ...globals[name],
              };
            }
          }
        } else {
          if (declaration.type === 'FunctionDeclaration') {
            const story = extractFunction(path, declaration);
            if (story) {
              const name = story.name;
              store.stories[name] = {
                ...story,
                ...globals[name],
              };
            }
          }
        }
      }
    },
  });
  if (store.doc) {
    store.doc.componentsLookup = components as Document['componentsLookup'];
  } else {
    throw new Error(`esm files should have one default export`);
  }
  return store;
};
