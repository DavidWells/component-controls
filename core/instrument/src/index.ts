import * as parser from '@babel/parser';
const mdx = require('@mdx-js/mdx');
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import prettier, { Options, ResolveConfigOptions } from 'prettier';
import parserBabel from 'prettier/parser-babylon';
import { StoriesGroup, Story } from '@component-controls/specification';
import { extractCSFStories } from './babel/csf-stories';
import { extractMDXStories } from './babel/mdx-stories';
import { removeMDXAttributes } from './babel/remove-mdx-attributes';

type TraverseFn = (stories: StoriesGroup) => any;

export type PrettierOptions = Options & {
  resolveConfigOptions?: ResolveConfigOptions;
};
const parseSource = async (
  code: string,
  traverseFn: TraverseFn,
  originalSource: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  let source: string;
  if (prettierOptions !== false) {
    const { resolveConfigOptions, ...otherOptions } = prettierOptions || {};
    let allPrettierOptions = otherOptions;
    if (filePath) {
      const userOptions = await prettier.resolveConfig(
        filePath,
        resolveConfigOptions,
      );
      allPrettierOptions = { ...userOptions, ...allPrettierOptions };
    }

    source = prettier.format(code, {
      parser: 'typescript',
      plugins: [parserBabel],
      ...allPrettierOptions,
    });
  } else {
    source = code;
  }

  const ast = parser.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  const stories = {
    stories: {},
    source: originalSource,
  };
  traverse(ast, traverseFn(stories));
  Object.keys(stories.stories).forEach((key: string) => {
    //@ts-ignore
    const story: Story = stories.stories[key];
    const { start, end } = story.loc || {};
    if (start && end) {
      const lines = source.split('\n');

      if (start.line === end.line) {
        story.source = lines[start.line - 1].substring(
          start.column,
          end.column,
        );
      } else {
        const startLine = lines[start.line - 1];
        const endLine = lines[end.line - 1];
        if (startLine !== undefined && endLine !== undefined) {
          story.source = [
            startLine.substring(start.column),
            ...lines.slice(start.line, end.line - 1),
            endLine.substring(0, end.column),
          ].join('\n');
        }
      }
    }
  });
  return stories;
};

export const parseCSF = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  return await parseSource(
    source,
    extractCSFStories,
    source,
    filePath,
    prettierOptions,
  );
};

export const parseMDX = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  const code = await mdx(source);

  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  traverse(ast, removeMDXAttributes());
  const newCode = generate(ast, {
    retainFunctionParens: true,
    retainLines: true,
  });
  return await parseSource(
    newCode.code,
    extractMDXStories,
    source,
    filePath,
    prettierOptions,
  );
};
