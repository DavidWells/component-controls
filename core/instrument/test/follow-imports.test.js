import { defaultParserOptions, defaultResolveOptions } from '../src/index';
import { followImports } from '../src/babel/follow-imports';

describe('follow-imports', () => {
  const extractFollowImportsForFile = (importName, fileName) => {
    return followImports(
      importName,
      require.resolve(fileName),
      defaultParserOptions,
      defaultResolveOptions,
    );
  };
  it('simple import', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/direct-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('one level import', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/one-level-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('one level default import', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/one-level-default-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('one level namespace import', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/one-level-namespace-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('one level index import', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/one-level-index-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('export all', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/export-all.js',
      ),
    ).toMatchSnapshot();
  });
  it('cjs named export', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/cjs-import.js',
      ),
    ).toMatchSnapshot();
  });
  it('export node modules', () => {
    expect(
      extractFollowImportsForFile(
        'Button',
        './examples/follow-imports/button-from-node-nodules.js',
      ),
    ).toMatchSnapshot();
  });
});
