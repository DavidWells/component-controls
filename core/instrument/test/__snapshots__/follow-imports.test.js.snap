// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`follow-imports one level default import 1`] = `
Object {
  "exportedAs": "default",
  "filePath": "/Users/atanasster/component-controls/core/instrument/test/examples/follow-imports/button-default-export.js",
  "loc": Object {
    "end": Object {
      "column": 24,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
}
`;

exports[`follow-imports one level import 1`] = `
Object {
  "exportedAs": "Button",
  "filePath": "/Users/atanasster/component-controls/core/instrument/test/examples/follow-imports/button-named-export.js",
  "loc": Object {
    "end": Object {
      "column": 30,
      "line": 1,
    },
    "start": Object {
      "column": 28,
      "line": 1,
    },
  },
}
`;

exports[`follow-imports one level index import 1`] = `
Object {
  "exportedAs": "Button",
  "filePath": "/Users/atanasster/component-controls/core/instrument/test/examples/follow-imports/button-named-export.js",
  "loc": Object {
    "end": Object {
      "column": 30,
      "line": 1,
    },
    "start": Object {
      "column": 28,
      "line": 1,
    },
  },
}
`;

exports[`follow-imports one level namespace import 1`] = `
Object {
  "exportedAs": "default",
  "filePath": "/Users/atanasster/component-controls/core/instrument/test/examples/follow-imports/button-default-export.js",
  "loc": Object {
    "end": Object {
      "column": 24,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
}
`;

exports[`follow-imports simple import 1`] = `
Object {
  "exportedAs": "Button",
  "filePath": "/Users/atanasster/component-controls/core/instrument/test/examples/follow-imports/direct-import.js",
  "loc": Object {
    "end": Object {
      "column": 30,
      "line": 1,
    },
    "start": Object {
      "column": 28,
      "line": 1,
    },
  },
}
`;
